File: weekly-contest-323-2500-2503/2501/Solution.py
from typing import *
from itertools import combinations

#Time limited exceed
# class Solution:
#     def valid(self, test):
#         test = list(test)
#         test.sort()
#         for i in range(0, len(test)-1):
#             if(test[i]*test[i] != test[i+1]):
#                 return False
#         return True

#     def longestSquareStreak(self, nums: List[int]) -> int:
#         if(len(nums) < 2):
#             return -1
#         maxLen = 0
#         for i in range(2, len(nums)+1):
#             testList = list(combinations(nums, i))
#             print(testList)
#             for test in testList:
#                 if(self.valid(test)):
#                     if(len(test) > maxLen):
#                         maxLen = len(test)
#         if(maxLen == 0):
#             return -1
#         return maxLen

# solution = Solution()
# print(solution.longestSquareStreak([2,4]))

    

File: weekly-contest-323-2500-2503/2501/Solution.java
import java.util.*; 
class Solution {
    public static int longestSquareStreak(int[] nums) {
        int maxLen = -1; 
        Arrays.sort(nums); 
        HashMap<Integer, Integer> hm = new HashMap<>();
        boolean[] visited = new boolean[nums.length]; //don't visit exist square twice 
        for(int i = 0; i < nums.length; i++){
            visited[i] = false;  
        }
        for(int i = 0; i < nums.length; i++){
            hm.put(nums[i], hm.getOrDefault(nums[i], 0) + 1); 
        }
        for(int i = 0; i < nums.length; i++){
            //prevent integer overflow
            if(visited[i] || nums[i] > 46340){
                continue; 
            }
            visited[i] = true; 
            int currLen = 1; 
            int square = nums[i]*nums[i]; 
            while(hm.getOrDefault(square, 0) != 0){
                if(square > 46340){
                    currLen++; 
                    break; 
                }
                currLen++; 
                square *= square; 
            }
            maxLen = (currLen > maxLen && currLen > 1) ? currLen : maxLen; 
        }
        return maxLen; 
    }
}
